torchdata/dataloader2/reading_service.py

```diff
@@ -102,24 +101,42 @@ def _collate_no_op(batch):
 class _IterateQueueDataPipes:
     def __init__(self, datapipes):
         self.datapipes = datapipes
+        for dp in self.datapipes:
+            if not isinstance(dp, communication.iter.QueueWrapper):
+                raise Exception("Source datapipes should be an instance of iter.QueueWrapper")
 
     def __iter__(self):
-        # TODO(612): This is slow as it does not sends data requests ahead.
-        exclude_datapipes: List[Any] = []
-        while len(exclude_datapipes) < len(self.datapipes):
-            for dp in self.datapipes:
-                if dp not in exclude_datapipes:
-                    forever = True
-                    while forever:
-                        try:
-                            value = dp.nonblocking_next()
-                            yield value
-                            forever = False
-                        except StopIteration:
-                            exclude_datapipes.append(dp)
-                            forever = False
-                        except communication.iter.NotAvailable:
-                            time.sleep(0.001)
+        self.reset()
+        total_pipes = len(self.datapipes)
+        disabled_pipe = [False] * len(self.datapipes)
+        cnt_disabled_pipes = 0
+
+        for idx in range(total_pipes):
+            self.datapipes[idx].protocol.request_next()
```

Comment:
Might be a noob question. We now request and receive data from `protocol` object. Then, do we still need `QueueWrapper`? We can directly let `_IterateQueueDataPipes` store a list of protocol clients.

Reply:
However, I'm still considering the possibility of merging _IterateQueueDataPipes and QueueWrapper to make it one class that supports 1:M queues.